### PR TITLE
Expand support for configurable menus

### PR DESCRIPTION
**Pull request by aleksip**
This PR builds on the current configurable account menu implementation, adding a new navigation plugin type for easier implementation of other configurable menus and more clearly separating the model from the view. The current account and admin menu implementations are refactored to use the new plugin type.

This is a first step of planned future improvement suggestions for the admin module and especially its navigation.

### CODE REVIEW

**Comment by demiankatz** [comment]
@aleksip, are you aware of #4006? At a glance, it looks like the work here and the work there are likely to conflict, and we should figure out a path forward....

**Comment by aleksip** [comment]
@demiankatz I was not. Thankfully it looks like the PRs touch some of the same files but implement different functionality. I'd be happy to revise this PR after #4006 is merged.

**Comment by demiankatz** [comment]
Thanks, @aleksip, I agree that it doesn't look like there should be major problems in resolving these two PRs... just a matter of whether ideas from #4006 should be incorporated here or vice versa! I think #4006 is nearing completion, but I'd be interested to see your review of that one (and, for that matter, @GrothTobias' review of this).

I'm going to hold off on closely reviewing this until things are a little more settled, but if you'd like me to begin digging in sooner rather than later, just let me know and I can be persuaded to make some time for it. :-)

**Comment by aleksip** [comment]
@demiankatz We do have plans for implementing groups like in #4006 and also nested menus. #4006 looks good to me in general. I could also incorporate the current changes made there into this PR if you'd rather have a single PR and it is ok with @GrothTobias. But I can also wait for #4006 to be merged and then revise this PR accordingly.

**Comment by demiankatz** [comment]
@aleksip, my expectation is that we can probably merge #4006 after one more round of revisions from @GrothTobias -- it's just a question of when he'll next have the availability to work on that. If that's going to be a blocker for you, I don't mind if you want to merge #4006 into this PR and make relevant adjustments. I expect it will be merged in a form very close to its present one. We can just put a TODO checkbox here reminding us to merge #4006 before merging this, to ensure that the separate phases of work remain separated.

**Comment by aleksip** [comment]
@demiankatz Sounds good to me! I think I won't have time for this before Nov 26 anyway actually.

**Comment by GrothTobias** [comment]
@demiankatz, if you want to keep commits to a minimum you can go ahead and merge #4006 into this one. 
@aleksip, do you plan to create a new PR for things like nested menus or add them in this PR? We could also try to implement them and maybe other improvements in #4006 if you want to keep these separated.

**Comment by aleksip** [comment]
@GrothTobias Nested menus would be a separate PR. They are on our roadmap for the next 3 months. If you are interested in implementing nested menus and need them sooner, please do go ahead! 🙂 In that case I'd be interested in reviewing the PR.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Okay, #4006 has been merged, so now it's time to resolve the conflicts here. Once that is done, I'll give this another closer review.

**Comment by aleksip** [comment]
@demiankatz I have now merged the latest commits in dev and refactored accordingly. I thought it was best not to touch the admin menu implementation more than necessary in this PR, so there is a note about options support in the new configuration file. It should be fully compatible with the new format and admin menu support for the new options could be added in a future PR.

File: config/vufind/AdminMenu.yaml
Comment (demiankatz) [inline config/vufind/AdminMenu.yaml:5]: Is there a reason for this? Do we expect these to eventually either reach parity or diverge significantly? (I'm asking mainly because I'm wondering if we could/should centralize more of the config documentation to prevent redundancy).

File: module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php]: Since this is a boolean method, would this be a better name?

```suggestion
    public function isContentActive()
```

File: module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php]: While refactoring anyway, might as well clean up this minor spacing issue...

```suggestion
            // assume that it is accessOnly
```

File: module/VuFind/src/VuFind/Navigation/AbstractMenu.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Navigation/AbstractMenu.php]: I think you could refactor this to `AccountMenu::getDefaultMenu()` to avoid needing a special case in the base class.

File: module/VuFind/src/VuFind/Navigation/AccountMenuFactory.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Navigation/AccountMenuFactory.php]: Probably worth null coalescing some defaults here... and maybe even calling `toArray()` above and using arrays, to help with our necessary migration away from Laminas\Config\Config.

Comment (demiankatz) [reply module/VuFind/src/VuFind/Navigation/AccountMenuFactory.php]: Also, why does the showOverdriveAdminMenu come into play here at all? Is this refactored from somewhere else? My understanding is that this setting should only impact the admin menu and not the account menu, but maybe I'm misunderstanding something.

File: themes/bootstrap5/templates/admin/menu.phtml
Comment (demiankatz) [inline themes/bootstrap5/templates/admin/menu.phtml:2]: Is it worth implementing menu groups here parallel to the myresearch menu? Would doing that allow you to share more code between the menu-related view helpers (i.e. reduce differentiated render methods)?

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @aleksip, this is looking pretty good to me -- just a few small suggestions.

It might also be worth considering adding some tests if time permits. Perhaps I can help with that at some point if you'd like me to!

File: config/vufind/AdminMenu.yaml
Comment (aleksip) [reply config/vufind/AdminMenu.yaml:5]: It probably would be a good thing if we could eventually refactor menu rendering to be based on the same templates as much as possible. And that would mean the options in all menu configuration files would be the same. But as this would be a big change I thought it might be good to go ahead in smaller steps.

File: module/VuFind/src/VuFind/Navigation/AccountMenuFactory.php
Comment (aleksip) [reply module/VuFind/src/VuFind/Navigation/AccountMenuFactory.php]: You are right, the code was copied from `OverdriveFactory` and `showOverdriveAdminMenu` is not relevant here.

File: themes/bootstrap5/templates/admin/menu.phtml
Comment (aleksip) [reply themes/bootstrap5/templates/admin/menu.phtml:2]: My thinking was that this is the model part of the refactoring and the view part would be a separate PR, especially as I am not sure how big a change that would be. But if you would prefer for everything to be done in one PR, I can give it a shot!

**Comment by aleksip** [comment]
@demiankatz I'd be happy to add some tests. Are there any suitable existing tests I could use as a basis?

File: config/vufind/AdminMenu.yaml
Comment (demiankatz) [reply config/vufind/AdminMenu.yaml:5]: As long as consistency is the end goal, and we don't forget to get there eventually, I agree that doing this work in smaller steps is a good idea, so the changes are easier to understand.

File: themes/bootstrap5/templates/admin/menu.phtml
Comment (demiankatz) [reply themes/bootstrap5/templates/admin/menu.phtml:2]: That makes sense; as I said above, I'm happy to keep this simple as long as we follow up with consistency as a next step.

**Comment by demiankatz** [comment]
> @demiankatz I'd be happy to add some tests. Are there any suitable existing tests I could use as a basis?

I think it depends on what kind of test you're interested in building. If you wanted to do Mink integration testing, you could use the [changeYamlConfigs](https://github.com/vufind-org/vufind/blob/dev/module/VuFind/src/VuFindTest/Integration/MinkTestCase.php#L205-L223) method of the default MinkTestCase base class to modify menu configurations and then confirm the results. If you want to do unit testing on some of the individual components you are building here, I think it would be pretty straightforward (apart from requiring a lot of mocking). Either or both of these approaches would be valuable and potentially helpful; just a question of what you think would be most feasible within your available time. And as noted, I'm happy to help (when time permits) if you want to share some of the load.

In any case, since I'm officially out of office this week, I can't devote much more time to this before next week... but I'm happy to continue discussing as I can! :-)

**Comment by aleksip** [comment]
@demiankatz I have now added unit tests for the new model classes. It probably is best to add Mink integration tests in the next step together with the template changes and new templates?

File: module/VuFind/src/VuFind/Navigation/AbstractMenu.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Navigation/AbstractMenu.php]: Shouldn't `$this->getDefaultMenuConfig()` become `static::getDefaultMenuConfig()` (here, and in other places where called) if we want to use a static method here?

File: module/VuFind/tests/unit-tests/src/VuFindTest/Navigation/AccountMenuTest.php
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Navigation/AccountMenuTest.php]: I would break this into two assertions so if there's a failure, the messaging is more clear:

```suggestion
        $this->assertCount(1, $menu['Account']['MenuItems']);
        $this->assertEquals('Profile', reset($menu['Account']['MenuItems'])['label']);
```

File: module/VuFind/tests/unit-tests/src/VuFindTest/Navigation/AccountMenuTest.php
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Navigation/AccountMenuTest.php]: You can use assertCount here:

```suggestion
        $this->assertCount(12, $menu['Account']['MenuItems']);
```

File: module/VuFind/tests/unit-tests/src/VuFindTest/Navigation/AdminMenuTest.php
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Navigation/AdminMenuTest.php]:
```suggestion
        $this->assertCount(6, $menu['Admin']['MenuItems']);
```

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @aleksip! A few minor suggestions based on the most recent commit.

**Review by demiankatz — APPROVED** [review]
Thanks, @aleksip, this looks good to me and all tests are passing.

**Comment by demiankatz** [comment]
@aleksip, I've added a [wiki page](https://vufind.org/wiki/development:plugins:navigation_helpers) to document the new helpers. Feel free to update (or let me know) if you see any errors/omissions.

**Comment by aleksip** [comment]
@demiankatz Thank you! The wiki page looks good, and I will keep it in mind for possible future updates.

**Pull request merged**
